app/integration_action.go

```diff
@@ -31,90 +31,152 @@ import (
 )
 
 func (a *App) DoPostAction(postId, actionId, userId, selectedOption string) (string, *model.AppError) {
+	return a.DoPostActionWithCookie(postId, actionId, userId, &model.DoPostActionRequest{
+		SelectedOption: selectedOption,
+	})
+}
+
+func (a *App) DoPostActionWithCookie(postId, actionId, userId string, actionRequest *model.DoPostActionRequest) (string, *model.AppError) {
+	// the prop values that we need to retain/clear in replacement message to match the original
+	remove := []string{"override_username", "override_icon_url"}
+	retain := map[string]interface{}{}
+
+	upstreamURL := ""
+	rootPostId := ""
+	upstreamRequest := &model.PostActionIntegrationRequest{
+		UserId: userId,
+		PostId: postId,
+	}
+
+	// See if the post exists in the DB, if so ignore the cookie.
+	// Start all queries here for parallel execution
 	pchan := a.Srv.Store.Post().GetSingle(postId)
 	cchan := a.Srv.Store.Channel().GetForPost(postId)
-
 	result := <-pchan
 	if result.Err != nil {
```

Comment:
(0/5) We should consider defaulting to the Cookie when available, and set it on all posts (with actions) sent to clients. It should be backwards-compatible. This would save 2 queries/DoPostAction, and with a few more changes may eventually eliminate the need to store message-specific PostActions in the database altogether.

Reply:
Agreed.